library.c

```diff
@@ -1779,6 +1783,63 @@ PHP_REDIS_API void redis_free_socket(RedisSock *redis_sock)
 }
 
 PHP_REDIS_API int
+redis_pack(RedisSock *redis_sock, zval *z, char **val, strlen_t *val_len TSRMLS_DC)
+{
+    char *buf, *data;
+    int valfree;
+    strlen_t len;
+    uint32_t res;
+
+    valfree = redis_serialize(redis_sock, z, &buf, &len TSRMLS_CC);
+    switch (redis_sock->compression) {
+        case REDIS_COMPRESSION_LZF:
+#ifdef HAVE_REDIS_LZF
+            data = emalloc(len);
+            res = lzf_compress(buf, len, data, len - 1);
+            if (res > 0 && res < len) {
+                if (valfree) efree(buf);
+                *val = data;
+                *val_len = res;
+                 return 1;
+            }
+            efree(data);
+#endif
+            break;
+    }
+    *val = buf;
+    *val_len = len;
+    return valfree;
+}
+
+PHP_REDIS_API int
+redis_unpack(RedisSock *redis_sock, const char *val, int val_len, zval *z_ret TSRMLS_DC)
+{
+    char *data;
+    int i, ret;
+    uint32_t res;
+
+    switch (redis_sock->compression) {
+        case REDIS_COMPRESSION_LZF:
+#ifdef HAVE_REDIS_LZF
+            errno = E2BIG;
+            for (i = 1; errno == E2BIG; ++i) {
+                data = emalloc(i * val_len);
+                if ((res = lzf_decompress(val, val_len, data, i * val_len)) == 0) {
+                    efree(data);
+                    continue;
```

Comment:
Rather `return 0;` ??

Reply:
No, increase buffer size and try again

Reply:
goto/continue are really bad for legibility...
BTW, doesn't return 0 is missing if res == 0 && erno != E2BUG.

Reply:
`res == 0 && err != E2BIG` breaks `for` loop

Reply:
Yes, but this is not reported to caller....

Reply:
Or perhaps you want to fallback on "uncompressed" data.... 
At least need some comments.

Reply:
I think we don't need to care about error here because  we return result of redis_unserialize. It will handle uncompressed values or return 0 if decompression fails

Reply:
`Or perhaps you want to fallback on "uncompressed" data....`
Exactly